File: go/path_sum_ii.go
package main

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func pathSum(root *TreeNode, targetSum int) [][]int {
	var result [][]int
	var walk func(*TreeNode, int, []int)
	walk = func(n *TreeNode, a int, v []int) {
		if n == nil {
			return
		}
		a += n.Val
		v = append(v, n.Val)
		if n.Left == nil && n.Right == nil && a == targetSum {
			w := make([]int, len(v))
			copy(w, v)
			result = append(result, w)
		}
		walk(n.Left, a, v)
		walk(n.Right, a, v)
	}
	walk(root, 0, nil)
	return result
}

File: c++/path_sum_ii.cpp
#include <bits/stdc++.h>

using namespace std;

struct TreeNode {
	int val;
	TreeNode *left;
	TreeNode *right;
	TreeNode() : val(0), left(nullptr), right(nullptr) {}
	TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
	TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution {
public:
	vector<vector<int>> pathSum(TreeNode *root, int targetSum) {
		vector<vector<int>> result;
		vector<int> path;
		findPaths(root, targetSum, path, result);
		return result;
	}

	void findPaths(TreeNode *n, int acc, vector<int> &path, vector<vector<int>> &result) {
		if (!n)
			return;
		acc -= n->val;
		path.push_back(n->val);
		if (!n->left && !n->right && acc == 0)
			result.push_back(vector<int>(path));
		findPaths(n->left, acc, path, result);
		findPaths(n->right, acc, path, result);
		path.pop_back();
	}
};
